pkg/audit/manager.go

```diff
@@ -185,47 +124,138 @@ func (am *auditManager) auditManagerLoop(ctx context.Context) {
-		return errors.Errorf("constraint %s status enforced not found", cname)
+	return discoveryClient.ServerResourcesForGroupVersion(constraintsGV)
+}
+
+func getUpdateListsFromAuditResponses(resp *constraintTypes.Responses) (map[string][]auditResult, error) {
+	updateLists := make(map[string][]auditResult)
+
+	for _, r := range resp.Results() {
+		selfLink := r.Constraint.GetSelfLink()
+		// skip if this constraint has reached the constraintViolationsLimit
+		if len(updateLists[selfLink]) == *constraintViolationsLimit {
+			continue
+		}
+
+		name := r.Constraint.GetName()
+		namespace := r.Constraint.GetNamespace()
+		apiVersion := r.Constraint.GetAPIVersion()
+		gvk := r.Constraint.GroupVersionKind()
+		message := r.Msg
+		if len(message) > msgSize {
+			message = truncateString(message, msgSize)
+		}
+		resource, ok := r.Resource.(*unstructured.Unstructured)
+		if !ok {
+			return nil, errors.Errorf("could not cast resource as reviewResource: %v", r.Resource)
+		}
+		rname := resource.GetName()
+		rkind := resource.GetKind()
+		rnamespace := resource.GetNamespace()
+		updateLists[selfLink] = append(updateLists[selfLink], auditResult{
+			cgvk:        gvk,
+			capiversion: apiVersion,
+			cname:       name,
+			cnamespace:  namespace,
+			rkind:       rkind,
+			rname:       rname,
+			rnamespace:  rnamespace,
+			message:     message,
+		})
 	}
-	if !val {
-		return errors.Errorf("constraint %s status not enforced", cname)
+	return updateLists, nil
+}
+
+func (am *AuditManager) updateConstraintsForKinds(ctx context.Context, resourceList *metav1.APIResourceList, updateLists map[string][]auditResult, timestamp string) error {
+	resourceGV := strings.Split(resourceList.GroupVersion, "/")
+	group := resourceGV[0]
+	version := resourceGV[1]
+
+	// get constraints for each Kind
+	for _, r := range resourceList.APIResources {
+		log.Info("constraint", "resource kind", r.Kind)
+		constraintGvk := schema.GroupVersionKind{
+			Group:   group,
+			Version: version,
+			Kind:    r.Kind,
+		}
+		instanceList := &unstructured.UnstructuredList{}
+		instanceList.SetGroupVersionKind(constraintGvk)
+		err := am.client.List(ctx, &client.ListOptions{}, instanceList)
+		if err != nil {
+			return err
+		}
+		log.Info("constraint", "count of constraints", len(instanceList.Items))
+		// get each constraint
+		for _, item := range instanceList.Items {
+			// if constraint is in updatedLists, then update its violations
+			constraintAuditResults, ok := updateLists[item.GetSelfLink()]
+			// else update auditTimestamp and clear violations
+			if !ok {
+				err = am.updateConstraintStatus(ctx, &item, emptyAuditResults, timestamp)
+				if err != nil {
+					return err
+				}
+			} else {
+				aResult := constraintAuditResults[0]
+				instance := &unstructured.Unstructured{}
+				instance.SetGroupVersionKind(aResult.cgvk)
+				namespacedName := types.NamespacedName{
+					Name:      aResult.cname,
+					Namespace: aResult.cnamespace,
+				}
+				// get the constraint
```

Comment:
We already have the constraint stored in `item`.

That being said the read/update of the constraint should probably be wrapped in some sort of retry framework.

I think the config finalizer cleanup subroutine has an example of the exponential backoff routine. We can put a context or a stop-channel on whatever retry framework to make sure we dont retry old commits after a new audit starts.